lib/schema.ts

```diff
@@ -0,0 +1,151 @@
+import {
+  AlevelQualification,
+  CommentType,
+  Decision,
+  DegreeCode,
+  FeeStatus,
+  GCSEQualification,
+  Gender,
+  NextAction,
+  Role
+} from '@prisma/client'
+import { formatISO } from 'date-fns'
+import { parse as parseDate } from 'date-fns/parse'
+import { ZodSchema, z } from 'zod'
+
+function numberSchema(from: number, to: number, fieldName: string, isNullable: boolean = false) {
+  let schema = z
+    .number()
+    .gte(from, { message: `${fieldName} must be ≥ ${from}` })
+    .lte(to, { message: `${fieldName} must be ≤ ${to}` })
+
+  return z.preprocess(
+    (val) => (val === '' ? null : Number(val)),
+    isNullable ? schema.nullable() : schema
+  )
+}
+
+const schemaCid = z.string().length(8, { message: 'CID must be exactly 8 characters' })
+const schemaAdmissionsCycle = z.coerce
+  .number()
+  .int()
+  .min(2020, { message: 'Admissions cycle must be in the 2020s or later' })
+
+export const schemaNextAction = z.nativeEnum(NextAction)
+
+// Schemas for forms
+export const schemaFormAdmin = z
+  .object({
+    gcseQualification: z.nativeEnum(GCSEQualification),
+    gcseQualificationScore: numberSchema(0, 10, 'Age 16 exam score'),
+    aLevelQualification: z.nativeEnum(AlevelQualification),
+    aLevelQualificationScore: numberSchema(0, 10, 'Age 18 exam score'),
+    motivationAdminScore: numberSchema(0, 10, 'Motivation score', true),
+    extracurricularAdminScore: numberSchema(0, 10, 'Extracurricular score', true),
+    examComments: z.string()
+  })
+  .partial()
+
+export const schemaFormReviewer = z.object({
+  motivationReviewerScore: numberSchema(0, 10, 'Motivation score'),
+  extracurricularReviewerScore: numberSchema(0, 10, 'Extracurricular score'),
+  referenceReviewerScore: numberSchema(0, 10, 'Reference score'),
+  academicComments: z.string()
+})
+
+export const schemaFormOutcome = z.object({
+  offerCode: z.string(),
+  offerText: z.string(),
+  decision: z.nativeEnum(Decision)
+})
+
+export const schemaFormComment = z.object({
+  text: z.string(),
+  authorLogin: z.string(),
+  type: z.nativeEnum(CommentType)
+})
+
+// ----------------------------
+// Schemas for CSV uploads
+
+// nested schema for upserting an applicant and creating a new application
+export const schemaCsvApplication = z.object({
+  applicant: z.object({
+    cid: schemaCid,
+    ucasNumber: z.string().length(10, { message: 'UCAS number must be exactly 10 digits' }),
+    gender: z.nativeEnum(Gender),
+    firstName: z.string(),
+    surname: z.string(),
+    preferredName: z.string().nullable(),
+    dateOfBirth: z
+      .string()
+      .transform((value) => formatISO(parseDate(value, 'dd/MM/yyyy', new Date()))),
```

Comment:
Ok with the sttring format. If this one and/or the one on line 97 recur across different files, could be worth extracting them as common constants

Reply:
These are the only two date formats and they're different so I'm inclined to leave these as they are